sdks/python/apache_beam/transforms/distribution_counter.pyx

```diff
@@ -0,0 +1,134 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=True
+
+cimport cython
+from libc.stdint cimport int64_t, INT64_MAX
+from libc.stdlib cimport malloc
+
+cdef extern from "stdlib.h" nogil:
+  void *memset(void *str, int c, size_t n)
+
+cdef int64_t* MAX_LONG_10_FOR_LEADING_ZEROS = [19, 18, 18, 18, 18, 17, 17, 17,
+                                               16, 16, 16, 15,15, 15, 15, 14,
+                                               14, 14, 13, 13, 13, 12, 12, 12,
+                                               12, 11, 11, 11, 10, 10, 10, 9, 9,
+                                               9, 9, 8, 8,8, 7, 7, 7, 6, 6, 6,
+                                               6, 5, 5, 5, 4, 4, 4, 3, 3, 3, 3,
+                                               2, 2, 2, 1, 1, 1, 0, 0, 0]
+
+cdef unsigned long long* POWER_TEN = [10e-1, 10e0, 10e1, 10e2, 10e3, 10e4, 10e5,
+                                      10e6, 10e7, 10e8, 10e9, 10e10, 10e11,
+                                      10e12, 10e13, 10e14, 10e15, 10e16, 10e17,
+                                      10e18]
+
+
+cdef inline bint compare_to(int64_t x, int64_t y):
+  """return the sign bit of x-y"""
+  if x < y:
+    return 1
+  return 0
+
+
+cdef int64_t bit_length(int64_t element):
+  """Same function as python built_in int.bit_length"""
+  cdef int64_t bit_count = 0
+  while element > 0:
+    bit_count += 1
+    element >>= 1
+  return bit_count
+
+
+cdef int64_t get_log10_round_to_floor(int64_t element):
+  cdef int64_t number_of_leading_zeros = 64 - bit_length(element)
+  cdef int64_t y = MAX_LONG_10_FOR_LEADING_ZEROS[number_of_leading_zeros]
+  return y - compare_to(element, POWER_TEN[y])
+
+
+cdef class DistributionAccumulator(object):
+  """Distribution Counter:
+  contains value distribution statistics and methods for incrementing
+  Attributes:
+    min: minimum value of all inputs.
+    max: maximum value of all inputs.
+    count: total count of all inputs.
+    sum: sum of all inputs.
+    first_bucket_offset: starting index of the first stored bucket.
+    last_bucket_offset: end index of buckets.
+    buckets: histogram buckets of value counts for a distribution
+                                                    (1,2,5 bucketing).
+             max bucket_index is 58( sys.maxint as input)
+    buckets_per_10: 3 buckets for every power of ten -> 1, 2, 5
+  """
+  def __init__(self):
+    self.min = INT64_MAX
+    self.max = 0
+    self.count = 0
+    self.sum = 0
+    self.first_bucket_offset = 58
+    self.last_bucket_offset = 0
+    self.buckets = <int64_t*> malloc(sizeof(int64_t) * 59)
```

Comment:
Can we clean up this memory?  (Right now, it is a leak.)

Reply:
Fixed in https://github.com/apache/beam/pull/5028/commits/38496c790993f0626d75f7afc2e54295fc80ec75